### PR TITLE
GH-100000: Cleanup and polish various watchers code

### PR DESCRIPTION
**Pull request by itamaro**
This includes:

1. Followup on @markshannon's [comment](https://github.com/python/cpython/pull/99859/files#r1038046404), improving code watchers dispatch efficiency (read callback only if watcher is on the active watchers bits, and bail out early from dispatch loop)
2. Apply same optimization to func watchers dispatch loop as well
3. Remove dict watchers array that was never used
4. Add initialization of type watchers array
5. Add `(PyCFunction)` cast to a few test functions
6. Change an invariant test in type watchers dispatch loop to an assert

* Issue: gh-100000

### CODE REVIEW

**Comment by itamaro** [comment]
I'm pretty sure this doesn't require a news entry, but not sure whether an issue is needed. let me know and I can file an issue if needed!

**Comment by itamaro** [comment]
filed an issue preemptively, got the nice and round GH-100000 ^_^

File: Objects/funcobject.c [hunk 1]
```diff
@@ -12,11 +12,18 @@ static void
...
-        PyFunction_WatchCallback cb = interp->func_watchers[i];
-        if ((cb != NULL) && (cb(event, func, new_value) < 0)) {
-            PyErr_WriteUnraisable((PyObject *) func);
+    uint8_t bits = interp->active_func_watchers;
+    int i = 0;
+    while (bits) {
+        assert(i < FUNC_MAX_WATCHERS);
+        if (bits & 1) {
+            PyFunction_WatchCallback cb = interp->func_watchers[i];
+            if ((cb != NULL) && (cb(event, func, new_value) < 0)) {
```
Comment (carljm) [inline Objects/funcobject.c]: you move the `cb != NULL` into an assert above for code watchers but not here, any reason for the inconsistency?

**Review by carljm — APPROVED** [review]
Looks good to me, thanks for the cleanup!

File: Objects/funcobject.c (hunk 1 above)
Comment (itamaro) [reply Objects/funcobject.c]: just an oversight! I'll change that to an assert too

**Comment by markshannon** [comment]
Looks good. Thanks.

**Pull request merged**